src/core/string_consumer.hpp

```diff
@@ -0,0 +1,889 @@
+ */
+class StringConsumer {
+public:
+	using size_type = std::string_view::size_type;
+
+	/**
+	 * Special value for "end of data".
+	 */
+	static constexpr size_type npos = std::string_view::npos;
+
+	/**
+	 * ASCII whitespace characters, excluding new-line.
+	 * Usable in FindChar(In|NotIn), (Peek|Read|Skip)(If|Until)Char(In|NotIn)
+	 */
+	static const std::string_view WHITESPACE_NO_NEWLINE;
+	/**
+	 * ASCII whitespace characters, including new-line.
+	 * Usable in FindChar(In|NotIn), (Peek|Read|Skip)(If|Until)Char(In|NotIn)
+	 */
+	static const std::string_view WHITESPACE_OR_NEWLINE;
+
+private:
+	std::string_view src;
+	size_type position = 0;
+
+	static void LogError(std::string &&msg);
+
+public:
+	/**
+	 * Construct parser with data from string.
+	 */
+	explicit StringConsumer(std::string_view src) : src(src) {}
+	/**
+	 * Construct parser with data from string.
+	 */
+	explicit StringConsumer(const std::string &src) : src(src) {}
+	/**
+	 * Construct parser with data from span.
+	 */
+	explicit StringConsumer(std::span<const char> src) : src(src.data(), src.size()) {}
+	/**
+	 * Construct parser with data from buffer.
+	 */
+	StringConsumer(const char* src, size_type len) : src(src, len) {}
+
+	/**
+	 * Check whether any bytes left to read.
+	 */
+	[[nodiscard]] bool AnyBytesLeft() const noexcept { return this->position <  this->src.size(); }
+	/**
+	 * Get number of bytes left to read.
+	 */
+	[[nodiscard]] size_type GetBytesLeft() const noexcept { return this->src.size() - this->position; }
+
+	/**
+	 * Check wheter any bytes were already read.
+	 */
+	[[nodiscard]] bool AnyBytesRead() const noexcept { return this->position > 0; }
+	/**
+	 * Get number of already read bytes.
+	 */
+	[[nodiscard]] size_type GetBytesRead() const noexcept { return this->position; }
+
+	/**
+	 * Get the original data, as passed to the constructor.
+	 */
+	[[nodiscard]] std::string_view GetOrigData() const noexcept { return this->src; }
+	/**
+	 * Get already read data.
+	 */
+	[[nodiscard]] std::string_view GetReadData() const noexcept { return this->src.substr(0, this->position); }
+	/**
+	 * Get data left to read.
+	 */
+	[[nodiscard]] std::string_view GetLeftData() const noexcept { return this->src.substr(this->position); }
+
+	/**
+	 * Discard all remaining data.
+	 */
+	void SkipAll() { this->position = this->src.size(); }
+
+	/**
+	 * Peek binary uint8.
+	 * @return Read integer, std::nullopt if not enough data.
+	 */
+	[[nodiscard]] std::optional<uint8_t> PeekUint8() const;
+	/**
+	 * Try to read binary uint8, and then advance reader.
+	 */
+	[[nodiscard]] std::optional<uint8_t> TryReadUint8()
+	{
+		auto value = this->PeekUint8();
+		if (value.has_value()) this->SkipUint8();
+		return value;
+	}
+	/**
+	 * Read binary uint8, and advance reader.
+	 * @param def Default value to return, if not enough data.
+	 * @return Read integer, 'def' if not enough data.
+	 */
+	[[nodiscard]] uint8_t ReadUint8(uint8_t def=0)
```

Comment:
```suggestion
	[[nodiscard]] uint8_t ReadUint8(uint8_t def = 0)
```
Too much Python lately?